### DISCUSSION TITLE
Possible to use multiple origins with setGlobalDispatcher?

### DISCUSSION THREAD

**Discussion by paulrobertlloyd** [nodejs/undici#1440] [Q&A]
I’d like to mock the following request:

```
GET https://server.example/micropub?q=source&url=https://website.example/post/1000
Authorization: Bearer xxxxxxxxx
Accept: application/json
```

This initial request can end up interacting with 3 different origins:

* Query is made to a `https://server.example` with a bearer token
* The bearer token is verified by another server, say `https://token-endpoint.example/token`
* If authorised, the request then fetches `https://website.example/post/1000`

When setting up a `MockAgent` (or `MockPool`) for this request, it appears as though I can only set one origin:

```js
const mockAgent = new MockAgent()

const mockPool = mockAgent.get('https://server.example')
mockPool.intercept({ path: '/micropub' }).reply(200, 'foo')
```

I guess I could use a RegEx for the origin name, but is that the only conceivable approach? This would mean avoiding using `path` values that conflict with each other, and makes it less clear which origin is being mocked.

I thought I might be able to use `setGlobalDispatcher()` multiple times, but looks like it’s only the agent set last that gets used.

**Comment by mcollina** [comment]
the `MockAgent` is built for this purpose:

```js
const mockAgent = new MockAgent()

const mockPool1 = mockAgent.get('https://server.example')
mockPool.intercept({ path: '/micropub' }).reply(200, 'foo')

const mockPool2 = mockAgent.get('https://token-endpoint.example')
mockPool.intercept({ path: '/token' }).reply(200, 'foobar')

const mockPool3 = mockAgent.get('https://website.example')
mockPool.intercept({ path: '/post/1000' }).reply(200, 'article')
```

**Reply by paulrobertlloyd** [reply]
Maybe it’s a factor of how I’ve chosen to set up my test helpers (with intercepts grouped by origin). I have something similar to the above already, and was wondering if I could then do something like this:

```js
import mockPool1 from '../mocks/mock-pool-1.js';
import mockPool2 from '../mocks/mock-pool-2.js';
import mockPool3 from '../mocks/mock-pool-3.js';

setGlobalDispatcher(mockPool1)
setGlobalDispatcher(mockPool2)
setGlobalDispatcher(mockPool3)

// Tests
…
```

Or possibly even something like this:

```js
import mockPool1 from '../mocks/mock-pool-1.js';
import mockPool2 from '../mocks/mock-pool-2.js';
import mockPool3 from '../mocks/mock-pool-3.js';

setGlobalDispatcher([mockPool1, mockPool2, mockPool3])

// Tests
…
```

Does that make sense?

**Reply by mcollina** [reply]
That's not possible. I think it might be possible to add a `mockAgent.set(url, mockPool)` to add multiple `MockPool` to the same `MockAgent`. Would you like to send a PR?
